include/picolibrary/state_machine.h

```diff
@@ -0,0 +1,858 @@
+     *
+     * \return true if state is the state event handler for the currently active state or
+     *         one of its superstates.
+     * \return false if state is neither the state event handler for the currently active
+     *         state nor one of its superstates.
+     */
+    auto is_in( State_Event_Handler_Reference state ) noexcept
+    {
+        Path path;
+
+        path.discover( *this, *m_current_state, state );
+
+        return path.is_complete();
+    }
+
+  private:
+    /**
+     * \brief State path.
+     */
+    class Path;
+
+    /**
+     * \brief Remove the state event handlers for the superstates that are common to a
+     *        pair of paths.
+     *
+     * \param[in] first The first of the two paths.
+     * \param[in] second The second of the two paths.
+     */
+    static void remove_common_superstates( Path & first, Path & second ) noexcept;
+
+    class Path {
+      public:
+        /**
+         * \brief Max state depth (not including the implicit "top" superstate).
+         */
+        static constexpr auto MAX_STATE_DEPTH = std::size_t{ 8 };
+
+        /**
+         * \brief Path storage.
+         */
+        using Storage = Fixed_Size_Array<State_Event_Handler_Pointer, MAX_STATE_DEPTH>;
+
+        /**
+         * \brief The number of elements in the path.
+         */
+        using Size = Storage::Size;
+
+        /**
+         * \brief A const path iterator.
+         */
+        using Const_Iterator = Storage::Const_Iterator;
+
+        /**
+         * \brief A const path reverse iterator.
+         */
+        using Const_Reverse_Iterator = Storage::Const_Reverse_Iterator;
+
+        /**
+         * \brief Constructor.
+         */
+        Path() noexcept = default;
+
+        Path( Path && ) = delete;
+
+        Path( Path const & ) = delete;
+
+        ~Path() noexcept = default;
+
+        auto operator=( Path && ) = delete;
+
+        auto operator=( Path const & ) = delete;
+
+        /**
+         * \brief Discover a path.
+         *
+         * \param[in] hsm The HSM for which a path is being discovered.
+         * \param[in] begin The state event handler for the state at the beginning of the
+         *            path.
+         * \param[in] end The state event handler for the state at the end of the path.
+         *
+         * If end is top, the path will cover the range [begin,top) and will be be
+         * considered to be complete.
+         * If end is a superstate of begin, the path will cover the range [begin,end) and
+         * will be considered to be complete. If end is not a superstate of begin, the
+         * path will cover the range [begin,top) and will not be considered to be
+         * complete.
+         */
+        void discover( HSM & hsm, State_Event_Handler_Reference begin, State_Event_Handler_Reference end ) noexcept
+        {
+            // #lizard forgives the length
+
+            m_is_complete  = end == top;
```

Comment:
```suggestion
            m_is_complete  = end == top;
```
Function length should be under the limit once this is removed.